src/command/lsp/mod.rs

```diff
@@ -0,0 +1,346 @@
+                }
+                CompositionEvent::Error(err) => {
+                    debug!("Composition failed: {err}");
+                    let message = format!("Composition failed to run: {err}",);
+                    let diagnostic = Diagnostic::new_simple(Range::default(), message);
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![diagnostic])
+                        .await;
+                }
+                CompositionEvent::SubgraphAdded(CompositionSubgraphAdded {
+                    name,
+                    schema_source,
+                }) => {
+                    debug!("Subgraph {} added", name);
+                    language_server.add_subgraph(name, schema_source).await;
+                }
+                CompositionEvent::SubgraphRemoved(CompositionSubgraphRemoved { name }) => {
+                    debug!("Subgraph {} removed", name);
+                    language_server.remove_subgraph(&name).await;
+                }
+            }
+        }
+        Ok::<(), StartCompositionError>(())
+    });
+    Ok(())
+}
+
+async fn create_composition_stream(
+    supergraph_config_path: Utf8PathBuf,
+    federation_version: Option<FederationVersion>,
+    client_config: StudioClientConfig,
+    lsp_opts: LspOpts,
+) -> Result<BoxStream<'static, CompositionEvent>, StartCompositionError> {
+    let fetch_remote_subgraphs_factory = MakeFetchRemoteSubgraphs::builder()
+        .studio_client_config(client_config.clone())
+        .profile(lsp_opts.plugin_opts.profile.clone())
+        .build();
+    let fetch_remote_subgraph_factory = MakeFetchRemoteSubgraph::builder()
+        .studio_client_config(client_config.clone())
+        .profile(lsp_opts.plugin_opts.profile.clone())
+        .build()
+        .boxed_clone();
+    let resolve_introspect_subgraph_factory =
+        MakeResolveIntrospectSubgraph::new(client_config.service()?).boxed_clone();
+
+    let composition_pipeline = CompositionPipeline::default()
+        .init(
+            &mut stdin(),
+            fetch_remote_subgraphs_factory,
+            Some(FileDescriptorType::File(supergraph_config_path.clone())),
+            None,
+        )
+        .await?
+        .resolve_federation_version(
+            resolve_introspect_subgraph_factory.clone(),
+            fetch_remote_subgraph_factory.clone(),
+            federation_version,
+        )
+        .await?
+        .install_supergraph_binary(
+            client_config.clone(),
+            None,
+            lsp_opts.plugin_opts.elv2_license_accepter,
+            lsp_opts.plugin_opts.skip_update,
+        )
+        .await?;
+
+    // Spin up Runner
+    Ok(composition_pipeline
+        .runner(
+            TokioCommand::default(),
+            FsReadFile::default(),
+            FsWriteFile::default(),
+            client_config.service()?,
+            fetch_remote_subgraph_factory.boxed_clone(),
+            1,
```

Comment:
worth being a const? what does this `1` represent, threads or something?

Reply:
So it's the retry interval for introspection, but this is actually fixed in the ROVER-245 PR because it gets turned into a command line option, which is a lot clearer anyway!